src/RealizationOrder.cpp

```diff
@@ -13,6 +16,155 @@ using std::set;
 using std::vector;
 using std::pair;
 
+namespace {
+
+class CheckLoopCarriedDependence : public IRVisitor {
+private:
+    const string &func;
+    const vector<Expr> &prev_args;
+
+    using IRVisitor::visit;
+
+    void visit(const Call *op) {
+        if ((op->call_type == Call::Halide) && (func == op->name)) {
+            internal_assert(!op->func.defined())
+                << "Func should not have been defined for a self-reference\n";
+            internal_assert(prev_args.size() == op->args.size())
+                << "Self-reference should have the same number of args as the original\n";
+            for (size_t i = 0; i < op->args.size(); i++) {
+                if (!equal(op->args[i], prev_args[i])) {
+                    debug(4) << "Self-reference of " << op->name << " with different args "
+                             << "from the LHS. Encounter loop-carried dependence.\n";
+                    result = false;
+                    return;
+                }
+            }
+        }
+    }
+
+public:
+    bool result;
+
+    CheckLoopCarriedDependence(const string &f, const vector<Expr> &prev_args)
+        : func(f), prev_args(prev_args), result(true) {}
+};
+
+// Check to see if there is a loop-carried dependence in a 'func' update definition
+// with the previous definition it's computed with.
+bool has_loop_carried_dependence(const string &func, const vector<Expr> &prev_args,
+                                 const Definition &update) {
+    internal_assert(prev_args.size() == update.args().size());
+    for (size_t i = 0; i < prev_args.size(); ++i) {
+        if (!equal(prev_args[i], update.args()[i])) {
+            return false;
+        }
+    }
+
+    for (size_t i = 0; i < update.values().size(); ++i) {
+        CheckLoopCarriedDependence c(func, prev_args);
+        update.values()[i].accept(&c);
+        if (c.result) {
+            return true;
+        }
+    }
+    return false;
+}
+
+void find_fused_groups_dfs(string current,
```

Comment:
`group` is populated recursively. I think it's cleaner to just pass it as reference rather returning it.